File: 2017/19/part_2/code.py
import sys
from enum import Enum

class Direction(Enum):
    UP = 1
    DOWN = 2
    LEFT = 3
    RIGHT = 4

def change(lines, i, j, dir, count):
    if dir == Direction.DOWN or dir == Direction.UP:
        if lines[i][j-1] == '-' or (lines[i][j-1] >= 'A' and lines[i][j-1] <= 'Z'):
            return i, j-1, Direction.LEFT, count
        elif lines[i][j+1] == '-' or (lines[i][j+1] >= 'A' and lines[i][j+1] <= 'Z'):
            return i, j+1, Direction.RIGHT, count
    elif dir == Direction.RIGHT or dir == Direction.LEFT:
        if lines[i-1][j] == '|' or (lines[i-1][j] >= 'A' and lines[i-1][j] <= 'Z'):
            return i-1, j, Direction.UP, count
        elif lines[i+1][j] == '|' or (lines[i+1][j] >= 'A' and lines[i+1][j] <= 'Z'):
            return i+1, j, Direction.DOWN, count
    return -1, -1, None, count

def process(lines, i, j, dir, count):
    if lines[i][j] == ' ':
        return -1, -1, None, count
    count += 1
    if lines[i][j] == '+':
        return change(lines, i, j, dir, count)
    if dir == Direction.DOWN:
        i += 1
    elif dir == Direction.RIGHT:
        j += 1
    elif dir == Direction.UP:
        i -= 1
    elif dir == Direction.LEFT:
        j -= 1
    return i, j, dir, count

def main(lines):
    lines = [line[:-1] for line in lines]
    i = j = 0
    j = lines[0].index('|')
    dir = Direction.DOWN
    count = 0 
    while True:
        if dir != None:
            i, j, dir, count = process(lines, i, j, dir, count)
        else:
            break
    print(count)

if __name__ == "__main__":
    main(open(sys.argv[1], 'r').readlines())

File: 2017/22/part_1/code.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <algorithm>
#include <vector>
#include <map>
#include <cmath>

enum direction {UP, LEFT, DOWN, RIGHT};

direction turn_left(direction current) {
  switch (current) {
    case UP:
      return LEFT;
    case LEFT:
      return DOWN;
    case DOWN:
      return RIGHT;
    case RIGHT:
      return UP;
    default:
      std::cout << "This should never be seen" << std::endl;
  }
}

direction turn_right(direction current) {
  switch (current) {
    case UP:
      return RIGHT;
    case LEFT:
      return UP;
    case DOWN:
      return LEFT;
    case RIGHT:
      return DOWN;
    default:
      std::cout << "This should never be seen" << std::endl;
  }
}

std::map<std::pair<int64_t, int64_t>, bool> get_grid(std::ifstream & input_file) {
  std::map<std::pair<int64_t, int64_t>, bool> grid;
  std::vector<std::string> input;
  std::string line;

  while(std::getline(input_file, line)) {
    input.push_back(line);
  }
  int64_t mid_way = std::floor(input.size()/2.0);
  int64_t size = input.size();
  for (int64_t i = 0; i < size; ++i) {
    for (int64_t j = 0; j < size; ++j) {
      if (input[i][j] == '#') {
        grid[std::make_pair(i - mid_way, j - mid_way)] = true;
      }
    }
  }
  return grid;
}

int main(int /*argc*/, char *argv[]) {
  size_t iters = std::atoll(argv[1]);
  std::ifstream input_file(argv[2], std::ifstream::in);
  auto grid = get_grid(input_file);

  size_t count = 0;
  int64_t i = 0, j = 0;
  direction dir = direction::UP;
  for (size_t t = 0; t < iters; ++t) {
    auto pair = std::make_pair(i, j);
    auto found = grid.find(pair);
    if (found != grid.end()) {
      dir = turn_right(dir);
      grid.erase(pair);
    } else {
      dir = turn_left(dir);
      grid[pair] = true;
      ++count;
    }
    switch (dir) {
      case UP:
        --i;
        break;
      case LEFT:
        --j;
        break;
      case DOWN:
        ++i;
        break;
      case RIGHT:
        ++j;
        break;
      default:
        std::cout << "NEVER" << std::endl;
    }
  }

  std::cout << count << std::endl;
  return 0;
}
